cli/crates/production-server/src/server/engine.rs

```diff
@@ -0,0 +1,147 @@
+            });
+
+            let (headers, stream) = encode_stream_response(ray_id.clone(), payload_stream, format).await;
+            return (headers, Body::from_stream(stream)).into_response();
+        }
+        _ => (),
+    }
+
+    let Some(session) = gateway.authorize(ctx.headers_as_map().into()).await else {
+        match (request, streaming_format) {
+            (BatchRequest::Single(_), None) => {
+                let response = gateway_v2::Response::unauthorized();
+
+                return (response.status, response.headers, response.bytes).into_response();
+            }
+            (BatchRequest::Single(_), Some(format)) => {
+                let (headers, stream) = encode_stream_response(
+                    ray_id,
+                    stream::once(async { engine_v2::Response::error("Unauthorized", []) }),
+                    format,
+                )
+                .await;
+
+                return (headers, axum::body::Body::from_stream(stream)).into_response();
+            }
+            (BatchRequest::Batch(requests), _) => {
+                let batch_response = BatchResponse::Batch(
+                    std::iter::repeat_with(gateway_v2::Response::unauthorized)
+                        .take(requests.len())
+                        .collect(),
+                );
+
+                return batch_response.into_response();
+            }
+        }
+    };
+
+    let response = match (request, streaming_format) {
+        (BatchRequest::Single(mut request), None) => {
+            request.ray_id = ctx.ray_id.clone();
+            BatchResponse::Single(session.execute(&ctx, request).await)
+        }
+        (BatchRequest::Single(mut request), Some(streaming_format)) => {
+            request.ray_id = ctx.ray_id.clone();
+
+            let (headers, stream) =
+                encode_stream_response(ray_id, session.execute_stream(request), streaming_format).await;
+
+            tokio::spawn(wait(receiver));
+
+            return (headers, axum::body::Body::from_stream(stream)).into_response();
+        }
+        (BatchRequest::Batch(requests), None) => {
+            let mut responses = Vec::with_capacity(requests.len());
+            for mut request in requests {
+                request.ray_id = ctx.ray_id.clone();
+                responses.push(session.clone().execute(&ctx, request).await)
+            }
+            BatchResponse::Batch(responses)
+        }
+        (BatchRequest::Batch(_), Some(_)) => {
+            unreachable!("should have been dealt with above")
+        }
+    };
+
+    tokio::spawn(wait(receiver));
+
+    response.into_response()
+}
+
+async fn wait(mut receiver: UnboundedReceiver<BoxFuture<'static, ()>>) {
+    // Wait simultaneously on everything immediately accessible
+    future::join_all(std::iter::from_fn(|| receiver.try_recv().ok())).await;
+
+    // Wait sequentially on the rest
+    while let Some(fut) = receiver.recv().await {
+        fut.await;
+    }
```

Comment:
I've never seen this pattern, what is the reasoning?

Reply:
Copied from dev. Ping @Finistere

Reply:
On CF for background tasks we need to use `wait_until` to ensure the worker doesn't stop just after returning the response. The current cache implementation directly relies on it and the `RequestContext` exposes this notion with a channel. So this function waits for all futures put into it in parallel first and then sequentially, as initial futures can themselves use `wait_until` in theory. But I'm removing that. We should just use `tokio::spawn` in native.